digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/csvDataImporter/TrafficSignCsvImporter.scala

```diff
@@ -89,20 +235,155 @@ class TrafficSignCsvImporter(roadLinkServiceImpl: RoadLinkService, eventBusImpl:
     }
   }
 
+  override def verifyData(parsedRow: ParsedProperties, user: User): ParsedCsv = {
+
+    def hasRoadName: Boolean = {
+      val optRoadName = getPropertyValueOption(parsedRow, "roadName").asInstanceOf[Option[String]]
+      optRoadName match {
+        case Some(roadName) if !roadName.trim.isEmpty => true
+        case _ => false
+      }
+    }
+
+    var errorMessages: (List[String], Seq[CsvAssetRowAndRoadLink]) = (List(), Seq())
+    val optTrafficSignType = getPropertyValueOption(parsedRow, "trafficSignType").asInstanceOf[Option[String]]
+
+    /* start date validations */
+    val temporaryDevices = Seq(4,5)
+    val optLifeCycle = getPropertyValueOption(parsedRow, "lifeCycle").asInstanceOf[Option[Int]].get
+    val optStartDate = getPropertyValueOption(parsedRow, "startDate").asInstanceOf[Option[String]]
+    val optEndDate = getPropertyValueOption(parsedRow, "endDate").asInstanceOf[Option[String]]
+    val isValidDate = optTrafficSignType match {
+      case Some(signType) if (!isBlank(optLifeCycle.toString) && temporaryDevices.contains(optLifeCycle)) || TrafficSignType.applyNewLawCode(signType) == RoadWorks =>
+        (optStartDate, optEndDate) match {
+          case (Some(startDate), Some(endDate)) =>
+            try {
+              val startDateFormat = dateFormatter.parseDateTime(startDate)
+              val endDateFormat = dateFormatter.parseDateTime(endDate)
+              endDateFormat.isAfter(startDateFormat) || endDateFormat.isEqual(startDateFormat)
+            } catch {
+              case _: Throwable => false
+            }
+          case (_, _) => false
+        }
+      case _ => true
+    }
+
+    if (!isValidDate) {
+      errorMessages = (errorMessages._1 ++ List("Invalid dates"), Seq())
+    }
+    /* end date validations */
+
+    /* start direction validations */
+    val optTrafficDirection = getPropertyValueOption(parsedRow, "bearing").asInstanceOf[Option[String]]
+    val directionValidator = optTrafficDirection match {
+      case Some(direction) if direction.trim.nonEmpty => true
+      case Some(_) if hasRoadName => true
+      case None if hasRoadName => true
+      case _ =>
+        errorMessages = (errorMessages._1 ++ List("Invalid traffic sign direction"), Seq())
+        false
+    }
+    /* end direction validations */
+
+    /* start additional panels type validations */
+    val minAdditionalPanels = 1
+    val maxAdditionalPanels = 3
```

Comment:
minAdditionalPanels  and maxAdditionalPanels  are declared in verifyData and generateBasePanelProperties functions.
Why not make those as attributes of class?

Reply:
Suggestion applied